File: day7/day7.py
from itertools import product
from functools import reduce

with open('real.txt') as f:
    lines = f.read().splitlines()

def split_target(line):
    target, rest = line.split(': ')
    rest = rest.split(' ')
    return int(target), [int(x) for x in rest]

def reducer(x, y):
    if y[0] == '+':
        return x + y[1]
    elif y[0] == '*':
        return x * y[1]
    else:
        return int(str(x) + str(y[1]))

def check_valid(target, rest, operators):
    num_operators = len(rest) - 1
    if num_operators == 0 and rest[0] != target:
        return False
    perms = product(operators, repeat=num_operators)
    for perm in perms:
        ops_val = zip(perm, rest[1:])
        out = reduce(reducer, ops_val, rest[0])
        if out == target:
            return True
            
    return False

def solve(lines, operators):
    res = 0
    for line in lines:
        target, rest = split_target(line)
        if check_valid(target, rest, operators):
            res += target
    return res

def part1(lines):
    return solve(lines, ['+', '*'])

def part2(lines):
    return solve(lines, ['+', '*', '||'])

print("part 1: ", part1(lines))
print("part 2: ", part2(lines))

File: day7/day7.clj
(ns day7)
(require 'helpers)


(defn parse_ops [ops values]
  (reduce (fn [acc [op val]] (
                              case op
                                "+" (+ acc val)
                                "*" (* acc val) 
                                "||" (Long/parseLong (str acc val))
                              )
                              ) 
          (first values) 
          (map vector ops (rest values)) 
  )
)

(defn is_valid? [target values perms]
  (if (empty? perms)
    false
    (if (= target (parse_ops (first perms) values))
      true
      (recur target values (rest perms)))
    )
)

(defn check_valid [[target values operators]]
  (let 
   [
    permutations 
    (helpers/cartesian_product (repeat (- (count values) 1) operators))
    ]
    (is_valid? target values permutations)
  )
)

(defn solve [input ops]
  (reduce (fn [total [target values]] 
            (
             if (check_valid (conj [target values] ops)) 
             (+ total target)
             total
             ) 
            )
            0
          input
          )
  )

(defn get_answers [filename]
  (let [inputs 
        (helpers/parse_input_colon_separated filename)]
  (println (str "Part 1: " (solve inputs ["+" "*"])))
  (println (str "Part 2: " (solve inputs ["*" "+" "||"])))
  )
)

(get_answers "day7/real.txt")
